WalletWasabi/Blockchain/TransactionBuilding/ChangelessTransactionCoinSelector.cs

```diff
@@ -24,13 +24,15 @@ public static class ChangelessTransactionCoinSelector
 		// target = target amount + output cost
 		long target = txOut.Value.Satoshi + feeRate.GetFee(txOut.ScriptPubKey.EstimateOutputVsize()).Satoshi;
 
-		// Keys are effective values of smart coins in satoshis.
-		IOrderedEnumerable<SmartCoin> sortedCoins = availableCoins.OrderByDescending(x => x.EffectiveValue(feeRate).Satoshi);
+		var coinsByScript = availableCoins
+			.GroupBy(coin => coin.ScriptPubKey.Hash)
+			.OrderByDescending(group => group.Sum(coin => coin.EffectiveValue(feeRate).Satoshi))
+			.ToList();
 
 		// How much it costs to spend each coin.
-		long[] inputCosts = sortedCoins.Select(x => feeRate.GetFee(x.ScriptPubKey.EstimateInputVsize()).Satoshi).ToArray();
+		long[] inputCosts = coinsByScript.Select(group => group.Sum(coin => feeRate.GetFee(coin.ScriptPubKey.EstimateInputVsize()).Satoshi)).ToArray();
 
-		Dictionary<SmartCoin, long> inputEffectiveValues = new(sortedCoins.ToDictionary(x => x, x => x.EffectiveValue(feeRate).Satoshi));
+		Dictionary<IEnumerable<SmartCoin>, long> inputEffectiveValues = new(coinsByScript.ToDictionary(x => x.Select(coin => coin), x => x.Sum(coin => coin.EffectiveValue(feeRate).Satoshi)));
```

Comment:
When calculating `inputCosts` and `inputEffectiveValues`, we consider address-reused coins as one coin, so we sum their values together to get one result to work with in _Branch and Bound_.

Reply:
Maybe leave commens in code for future reviewing.